### PR TITLE
Implement <latitude-embed> webcomponent

### PR DESCRIPTION
**Pull request by andresgutgon**
# What?
Following [this PR](https://github.com/latitude-dev/latitude/pull/241) we have now the boilerplate in place for doing a web component for interacting with a latitude project inside an iframe from a React or VanillaJS project

### TODO
- [x] `latitude-embed` has to be responsive. When the host size change inner iframe has to be updated.
- [x] Allow host to run all or some queries present in the latitude view
- [x] Allow host to update one or more params and the iframe's URL has to be changed also internal state of those params 
- [x] Allow to pass and update `signedToken` and update internal iframe URL
- [x] Allow host to listen to params changes inside the iframe

### TODO
- [ ] Check everything works with published packages
- [ ] Add documentation

### CODE REVIEW

File: apps/server/src/lib/iframeEmbedding.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  }
+  iframe.onRun = (data) => {
+    computeQueries(data)
+  }
+
+  useViewParams().subscribe((params) => {
+    iframe.sendParamsChanged(params)
+  })
+
+  iframe.listen()
```
Comment (andresgutgon) [inline apps/server/src/lib/iframeEmbedding.ts:22]: Start listening to `<iframe />` messages. If the page sending the messages is in the `allowedOrigins` the message will arrive. 

Note the FIXME here. For now this check is fake and we allow any page to embed any latitude project

File: examples/sample-react/src/examples/Embedding/index.tsx [hunk 2]
```diff
@@ -1,27 +1,103 @@
...
-  const [lastName, setLastName] = useState('Merlo')
+  const starwarsFetched = useRef(false)
+  const [characters, setCharacters] = useState<Option[]>([])
+  const { headerHeight } = useExample()
+  const [ref, { height }] = useMeasure<HTMLDivElement>()
+  const topHeight = height + headerHeight
+  const [endYear, setEndYear] = useState(2006)
+
+  const runQuery = useCallback(() => {
+    runEmbedViewQuery({ queryPaths: [], force: true })
```
Comment (andresgutgon) [inline examples/sample-react/src/examples/Embedding/index.tsx:30]: Run all queries (empty list) or some specific query in the iframe from react with this `runEmbedViewQuery` method

File: examples/sample-react/src/examples/Embedding/index.tsx [hunk 3]
```diff
@@ -1,27 +1,103 @@
...
+        <div className='absolute -top-2.5 left-4 px-2 py-0.5 text-xs text-blue-700 uppercase rounded border border-blue-500 bg-white'>Your React app</div>
+        <Button onClick={runQuery}>Run</Button>
+        <div className='flex-grow-0'>
+          <Input
+            type='number'
+            name='name'
+            value={endYear}
+            onChange={(e) => {
+              const value = Number(e.target.value)
+              changeEmbedParams({ end_year: value })
```
Comment (andresgutgon) [inline examples/sample-react/src/examples/Embedding/index.tsx:68]: When a param change in the react app notify the iframe

File: examples/sample-react/src/examples/Embedding/index.tsx [hunk 4]
```diff
@@ -1,27 +1,103 @@
...
+              setEndYear(value)
+            }}
+          />
+        </div>
+        <Combobox
+          options={characters}
+          onChange={(selected) => {
+            triggerCustomEvent(selected)
```
[the remark this answers is not in the document]
Comment (andresgutgon) [reply examples/sample-react/src/examples/Embedding/index.tsx:76]: In this case when we receive the event we fetch from SWAPI the character

``` javascript
<script>
  import { onMount } from 'svelte'
  import { writable } from 'svelte/store'
  import { iframe } from '@latitude-data/embedding'

  function buildImageSrc(data) {
    const url = data?.url
    if (!url) return null
    const id = data.url.split('/').filter(Boolean).pop()
    return `https://raw.githubusercontent.com/vieraboschkova/swapi-gallery/main/static/assets/img/people/${id}.jpg`
  }

  const characterData = writable(undefined)
  const imageUrl = writable(undefined)

  onMount(async () => {
    iframe.onCustomEvent = async (event) => {
      const data = event.data
      if (!data) return

      try {
        const response = await window.fetch(data.value);
        if (response.ok) {
          const character = await response.json()
          characterData.set(character)
          const src = buildImageSrc(character)
          imageUrl.set(src)
        } else {
          console.error("Failed to fetch data:", response.status)
        }
      } catch (error) {
        console.error("Error fetching data:", error)
      }
    }
  })
</script>

<svelte:head>
  <title>Netflix Analysis</title>
</svelte:head>
```

Note the `iframe.onCustomEvent` callback.

File: packages/client/webcomponents/src/components.d.ts [hunk 5]
```diff
@@ -5,15 +5,32 @@
...
+export interface LatitudeEmbedCustomEvent<T> extends CustomEvent<T> {
+    detail: T;
+    target: HTMLLatitudeEmbedElement;
+}
 declare global {
+    interface HTMLLatitudeEmbedElementEventMap {
+        "paramsChanged": EmbeddingEventData<EmbeddingEvent.ParamsChanged>;
+    }
     interface HTMLLatitudeEmbedElement extends Components.LatitudeEmbed, HTMLStencilElement {
+        addEventListener<K extends keyof HTMLLatitudeEmbedElementEventMap>(type: K, listener: (this: HTMLLatitudeEmbedElement, ev: LatitudeEmbedCustomEvent<HTMLLatitudeEmbedElementEventMap[K]>) => any, options?: boolean | AddEventListenerOptions): void;
```
Comment (andresgutgon) [inline packages/client/webcomponents/src/components.d.ts:26]: Auto generated types for the component are not working. I'll left a FIXME

File: packages/client/webcomponents/src/components/latitude-embed/latitude-embed.tsx
Comment (andresgutgon) [inline packages/client/webcomponents/src/components/latitude-embed/latitude-embed.tsx:1]: All the logic for the WebComponent (React) is here

File: apps/server/src/lib/stores/viewParams.ts [hunk 6]
```diff
@@ -19,55 +19,57 @@ const getParamsFromUrl = () => {
...
+
 const viewParamsStore = writable<ViewParams>(getParamsFromUrl())
 
 export const useViewParams = (): Writable<ViewParams> => viewParamsStore
 export const getAllViewParams = (): ViewParams => get(viewParamsStore)
 
+useViewParams().subscribe((params) => {
+  setUrlParam(params)
+})
+
```
Comment (csansoon) [inline apps/server/src/lib/stores/viewParams.ts:47]: Could this be ran multiple times within the same client? If so, that may affect performance

File: apps/server/src/routes/+layout.svelte [hunk 7]
```diff
@@ -15,7 +17,16 @@
...
-  onMount(initQueries)
+  onMount(async () => {
+    await initQueries()
+
+    const iframe = initIframeCommunication({ allowedOrigins: ['*'] })
+
+    useViewParams().subscribe((params) => {
+      setUrlParam(params)
+      iframe.sendParamsChanged(params)
+    })
```
Comment (andresgutgon) [inline apps/server/src/routes/+layout.svelte:28]: Everything related with params changing in a latitude view happens here.

1. URL is updated
2. iframes are notified

File: apps/server/src/lib/stores/queries.ts [hunk 8]
```diff
@@ -225,20 +225,25 @@ export async function computeQueries({
...
+      )
+      .map((queryInView) =>
+        fetchQueryFromCore({
+          query: queryInView.queryPath,
+          inlineParams: queryInView.inlineParams,
+          force,
+          skipIfParamsUnchanged,
+        }),
+      ),
+  )
```
Comment (csansoon) [inline apps/server/src/lib/stores/queries.ts:246]: I like this

File: packages/client/webcomponents/src/components.d.ts [hunk 9]
```diff
@@ -25,9 +42,10 @@ declare global {
...
 declare namespace LocalJSX {
     interface LatitudeEmbed {
-        "first"?: string;
-        "last"?: string;
-        "middle"?: string;
+        "onParamsChanged"?: (event: LatitudeEmbedCustomEvent<EmbeddingEventData<EmbeddingEvent.ParamsChanged>>) => void;
+        "params"?: Record<string, string>;
+        "query": string;
+        "signedParams"?: string;
     }
```
Comment (csansoon) [inline packages/client/webcomponents/src/components.d.ts:49]: Why is `query` needed on a `LatitudeEmbed`?

**Review by csansoon — DISMISSED** [review]

File: .changeset/dirty-nails-lie.md [hunk 10]
```diff
@@ -0,0 +1,9 @@
+---
+"@latitude-data/webcomponents": minor
+"@latitude-data/react": minor
+"@latitude-data/client": minor
+"@latitude-data/embedding": minor
+"@latitude-data/server": minor
+---
+
+Improve embedding experience bettween users apps and Latitude iframe. We did a integration as a webcomponent that users can include in their projects and listen to changes in the params inside Latitude's view. Also they can trigger view recomputations
```
Comment (csansoon) [inline .changeset/dirty-nails-lie.md:9]: between*

**Review by geclos — APPROVED** [review]

**Pull request merged**